File: LeetCode - May/permutationInString.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        
        if(s1.length() > s2.length()) 
			return false;
        
        int[] s1hash = new int[26];
        int[] s2hash = new int[26];
        
        for(int i = 0; i < s1.length(); i++){
            s1hash[s1.charAt(i) - 'a']++;
            s2hash[s2.charAt(i) - 'a']++;
        }
        
        if(Arrays.equals(s1hash, s2hash)) 
			return true;
        
        int start = 0;
        int end = s1.length();
        
		while(end < s2.length()){
            s2hash[s2.charAt(start) - 'a']--;
            s2hash[s2.charAt(end) - 'a']++;
            
            if(Arrays.equals(s1hash, s2hash)) 
				return true;
            start++;
            end++;
        }
        return false;      

        
        
    }
} 

File: LeetCode - May/permutationInString.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if(len(s1)>len(s2)):
            return False
        
        s1hash = [0 for i in range(26)]
        s2hash = [0 for i in range(26)]
        
        for i in range(len(s1)):
            s1hash[ord(s1[i])-ord('a')]+=1
            s2hash[ord(s2[i])-ord('a')]+=1
        
        if(s1hash == s2hash):
            return True
        
        start=0
        end=len(s1)
        while(end<len(s2)):
            
            s2hash[ord(s2[start])-ord('a')]-=1
            s2hash[ord(s2[end])-ord('a')]+=1
            
            if( s1hash == s2hash ):
                return True
            
            start+=1
            end+=1
        return False